packages/graphql/src/registry.ts

```diff
@@ -0,0 +1,46 @@
+import { navigateProgram, type Program, type SemanticNodeListener } from "@typespec/compiler";
+import type { GraphQLObjectType } from "graphql";
+
+type Mutable<T> = {
+  -readonly [k in keyof T]: T[k];
+};
+
+// This class contains the registry of all the GraphQL types that are being used in the program
+export class GraphQLTypeRegistry {
+  program: Program;
+  readonly programNavigated: boolean = false;
+
+  constructor(program: Program) {
+    this.program = program;
+    return new Proxy(this, {
+      get(target: GraphQLTypeRegistry, prop: string, receiver) {
+        if (GraphQLTypeRegistry.#publicGetters.includes(prop)) {
+          if (!target.programNavigated) {
+            const mutableThis = target as Mutable<GraphQLTypeRegistry>;
+            navigateProgram(target.program, target.#semanticNodeListener);
+            mutableThis.programNavigated = true;
+          }
+        }
+        return Reflect.get(target, prop, receiver);
```

Comment:
I haven't seen the use of reflection in the TSP repo. I am not sure that this is good typescript practice. I am not an expert on this, but in compiled languages at least using reflection is usually a code smell. Besides, I am not quite convinced that this is needed as per my comment below.

Reply:
`Reflect` in JavaScript isn't really the same concept as reflection in other languages:
> The major use case of Reflect is to provide default forwarding behavior in Proxy handler traps.
— [JS docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Reflect#description)

The way we're using it here is how it's expected to be used — when you are creating a `Proxy` object, it's used to call through to the original proxied object.

Reply:
Also, this actually _is_ used [in the TypeSpec repo](https://github.com/microsoft/typespec/blob/f89235377963cf5ad0da6f1ef4d0e2c3035dce1b/packages/compiler/src/experimental/typekit/define-kit.ts#L28-L49). It's in the "experimental" `TypeKit` stuff, which is the new code that Timothee seems to have been working on as the replacement for the existing emitter framework. Since it's some of the newest code, I imagine it's up to the standards that the Microsoft team intends to use.